File: py_validation/pardox/_bindings/sql_query.py
def bind(lib, c_void_p, c_char_p, c_int32, c_double, c_longlong, c_size_t, c_int64, c_int16):
    if hasattr(lib, 'pardox_sql_query'):
        lib.pardox_sql_query.argtypes = [c_void_p, c_char_p]
        lib.pardox_sql_query.restype = c_void_p

    if hasattr(lib, 'pardox_execute_sql'):
        lib.pardox_execute_sql.argtypes = [c_char_p, c_char_p]
        lib.pardox_execute_sql.restype = c_longlong

    if hasattr(lib, 'pardox_write_sql'):
        lib.pardox_write_sql.argtypes = [c_void_p, c_char_p, c_char_p, c_char_p, c_char_p]
        lib.pardox_write_sql.restype = c_longlong

    if hasattr(lib, 'pardox_write_sql_prdx'):
        lib.pardox_write_sql_prdx.argtypes = [c_char_p, c_char_p, c_char_p, c_char_p, c_char_p, c_longlong]
        lib.pardox_write_sql_prdx.restype = c_longlong

    if hasattr(lib, 'pardox_sqlserver_config_ok'):
        lib.pardox_sqlserver_config_ok.argtypes = [c_char_p]
        lib.pardox_sqlserver_config_ok.restype = c_int32

    # Gap 30 — SQL Cursor API (streaming batch reads from SQL databases)
    if hasattr(lib, 'pardox_scan_sql_cursor_open'):
        lib.pardox_scan_sql_cursor_open.argtypes = [c_char_p, c_char_p, c_size_t]
        lib.pardox_scan_sql_cursor_open.restype = c_void_p

    if hasattr(lib, 'pardox_scan_sql_cursor_fetch'):
        lib.pardox_scan_sql_cursor_fetch.argtypes = [c_void_p]
        lib.pardox_scan_sql_cursor_fetch.restype = c_void_p

    if hasattr(lib, 'pardox_scan_sql_cursor_offset'):
        lib.pardox_scan_sql_cursor_offset.argtypes = [c_void_p]
        lib.pardox_scan_sql_cursor_offset.restype = c_size_t

    if hasattr(lib, 'pardox_scan_sql_cursor_close'):
        lib.pardox_scan_sql_cursor_close.argtypes = [c_void_p]
        lib.pardox_scan_sql_cursor_close.restype = None

    if hasattr(lib, 'pardox_scan_sql_to_parquet'):
        lib.pardox_scan_sql_to_parquet.argtypes = [c_char_p, c_char_p, c_char_p, c_longlong]
        lib.pardox_scan_sql_to_parquet.restype = c_longlong

File: php_validation/pardox/src/Core/_bindings/sql_query.php
<?php
// Core/_bindings/sql_query.php — Gap 14: SQL query over DataFrame, write SQL, execute SQL
return <<<'CDECL'

            // --- Gap 14: SQL Query over in-memory DataFrame ---
            HyperBlockManager* pardox_sql_query(HyperBlockManager* mgr, const char* sql);

            // --- Writers ---
            long long pardox_write_sql(HyperBlockManager* mgr, const char* conn_str, const char* table, const char* mode, const char* conflict_cols_json);
            long long pardox_execute_sql(const char* conn_str, const char* query);

            // --- Gap 14: Streaming PRDX to SQL ---
            long long pardox_write_sql_prdx(const char* prdx_path, const char* conn_str, const char* table_name, const char* mode, const char* conflict_cols_json, long long batch_rows);

            // --- SQL Server Config Check ---
            int pardox_sqlserver_config_ok();

            // --- Gap 30: SQL Cursor API (streaming batch reads from SQL databases) ---
            void* pardox_scan_sql_cursor_open(const char* conn_str, const char* query, size_t chunk_size);
            void* pardox_scan_sql_cursor_fetch(void* cursor);
            size_t pardox_scan_sql_cursor_offset(void* cursor);
            void pardox_scan_sql_cursor_close(void* cursor);
            long long pardox_scan_sql_to_parquet(const char* conn_str, const char* query, const char* output_pattern, long long chunk_size);

CDECL;
